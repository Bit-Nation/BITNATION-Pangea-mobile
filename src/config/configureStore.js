// @flow

import { createStore, applyMiddleware, compose, type Store } from 'redux';
import logger from 'redux-logger';
import createSagaMiddleware from 'redux-saga';
import createSagaMonitor from '@clarketm/saga-monitor';
import reducers from '../reducers';
import rootSaga from '../sagas';

/**
 * @desc Configures a Redux store.
 * @return {Store} Created store object.
 */
export default function configureStore(): Store {
  const sagaMonitor = createSagaMonitor({
    level: 'log',
    actionDispatch: true,
  });
  const sagaMiddleware = createSagaMiddleware({ sagaMonitor });

  const enhancer = compose(
    applyMiddleware(sagaMiddleware),
    applyMiddleware(logger),
  );

  const store = createStore(reducers, enhancer);
  sagaMiddleware.run(rootSaga);
  return store;
}
