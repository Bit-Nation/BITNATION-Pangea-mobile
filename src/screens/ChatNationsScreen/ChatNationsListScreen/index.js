// @flow

import React from 'react';
import {
  View,
  SectionList,
  Clipboard,
} from 'react-native';
import _ from 'lodash';
import { Fab, Text } from 'native-base';
import ActionSheet from 'react-native-actionsheet';

import BackgroundImage from '../../../components/common/BackgroundImage';
import styles from './styles';
import { screen } from '../../../global/Screens';
import ChatListItem from '../../../components/common/ChatListItem';
import NationListHeader from '../../../components/common/NationListHeader';
import FakeNavigationBar from '../../../components/common/FakeNavigationBar';
import i18n from '../../../global/i18n';
import AssetsImages from '../../../global/AssetsImages';
import type { NationIdType, NationType } from '../../../types/Nation';
import type { Navigator } from '../../../types/ReactNativeNavigation';
import ScreenTitle from '../../../components/common/ScreenTitle';

type Props = {
  /**
   * @desc React Native Navigation navigator object.
   */
  navigator: Navigator,
  /**
   * @desc Selected Tab Name
   */
  selectedTab: string,
  /**
   * @desc List of all nations
   */
  nations: Array<NationType>,
  /**
   * @desc List of all nations
   */
  myNations: Array<NationType>,
  /**
   * @desc Function to be called when an item is selected from the list
   * @param id ID of the nation to be opened
   * @param isBot Flag to indicate if the selected channel is bot
   */
  onSelectItem: (id: number, isBot: boolean) => void,
};

const ChatNationsListScreen = ({
  selectedTab,
  nations,
  myNations,
  onSelectItem,
}: Props) => {
  const nationsToDisplay = selectedTab === 'ALL_NATIONS' ?
    _.filter(nations, nation => nation.idInSmartContract >= 0)
    :
    _.filter(nations, nation => (
      _.indexOf(myNations, nation.id) !== -1) && (nation.idInSmartContract >= 0));
  const sortedNations = _.sortBy(nationsToDisplay, nation => nation.nationName);
  const groups = _.groupBy(sortedNations, nation => nation.nationName.charAt(0));
  let sections = _.map(groups, (group, key) => ({
    title: key,
    data: group,
  }));
  const bots = [{ title: 'Bots', data: [{ name: 'Dr. FreudBot', isBot: true, id: 0 }] }];
  sections = bots.concat(sections);

  return (
    <View style={styles.nationsScreenContainer}>
      <BackgroundImage />
      <FakeNavigationBar />
      <ScreenTitle title={i18n.t('screens.chat.title')} />
      <SectionList
        renderItem={(item) => {
          const nation = item.item;
          if (nation.isBot === true) {
            return (<ChatListItem
              text={nation.name}
              participants=''
              itemIcon={AssetsImages.ChatUI.botIcon}
              onPress={id => onSelectItem(id, true)}
              id={nation.id}
            />);
          }
            return (<ChatListItem
              text={nation.nationName}
              participants=''
              itemIcon={0}
              onPress={id => onSelectItem(id, false)}
              id={nation.id}
            />);
        }}
        keyExtractor={item => item.id}
        renderSectionHeader={({ section }) => <NationListHeader title={section.title} />}
        sections={sections}
        style={styles.sectionList}
      />
    </View>
  );
};

export default ChatNationsListScreen;
