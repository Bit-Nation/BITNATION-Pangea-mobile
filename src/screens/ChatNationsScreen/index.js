// @flow

import React from 'react';
import { connect } from 'react-redux';

import ChatNationsListScreen from './ChatNationsListScreen';
import { switchNationTab, openNation } from '../../actions/nations';
import { screen } from '../../global/Screens';
import { resolveNation } from '../../utils/nations';
import NavigatorComponent from '../../components/common/NavigatorComponent';
import { startNationCreation } from '../../actions/modifyNation';
import type { Navigator } from '../../types/ReactNativeNavigation';
import type { NationIdType, NationType } from '../../types/Nation';

type Props = {
  /**
   * @desc React Native Navigation navigator object.
   */
  navigator: Navigator,
  /**
   * @desc Selected Tab Name
   */
  selectedTab: string,
  /**
   * @desc List of nations
   */
  nations?: Array<NationType>,
  /**
   * @desc List of nations ids that the current user has joined to.
   */
  myNationIds: Array<NationIdType>,
  /**
   * @desc Function to retrieve nations from the database
   */
  fetchNations: () => void,
  /**
   * @desc Function to open a nation
   * @param id Id of the nation to be opened
   */
  openNation: (id: NationIdType) => void,
};

class ChatNationsScreen extends NavigatorComponent<Props> {
  constructor(props: Props) {
    super(props);

    this.props.navigator.setButtons({
      leftButtons: [],
      rightButtons: [],
    });
  }

  render() {
    return (
      <ChatNationsListScreen onSelectItem={this.onSelectItem} {...this.props} />
    );
  }

  onSelectItem = (id, isBot) => {
    const nation = resolveNation(this.props.nations || [], id);

    if (!nation) {
      if (isBot === true) {
        this.props.navigator.push({
          ...screen('CHAT_SCREEN'),
          passProps: { isBot },
        });
      }
      return;
    }

    this.props.openNation(id);

    this.props.navigator.push({
      ...screen('CHAT_SCREEN'),
      passProps: { isBot },
    });
  };
}

const mapStateToProps = state => ({
  ...state.nations,
  ...state.wallet,
});

const mapDispatchToProps = dispatch => ({
  onSelectTab(index) {
    dispatch(switchNationTab(index === 0 ? 'ALL_NATIONS' : 'MY_NATIONS'));
  },
  openNation(id) {
    dispatch(openNation(id));
  },
  startNationCreation() {
    dispatch(startNationCreation());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ChatNationsScreen);
