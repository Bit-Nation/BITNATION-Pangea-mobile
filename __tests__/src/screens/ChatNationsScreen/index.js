import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import ChatNationsScreen from '../../../../src/screens/ChatNationsScreen';
import navigatorMock from '../../../../__mocks__/Navigator';

describe('ChatNationsScreen', () => {
  let chatScreen;

  const initialStateMock = {};
  const storeMock = configureStore([]);
  const propsMock = {
    navigator: navigatorMock,
    onSelectTab: jest.fn(),
    openNation: jest.fn(),
    fetchNations: jest.fn(),
    startNationCreation: jest.fn(),
  };

  beforeEach(() => {
    chatScreen = shallow(<ChatNationsScreen {...propsMock} store={storeMock(initialStateMock)} />);
  });

  test('Rendering', () => {
    expect(chatScreen).toMatchSnapshot();
  });

  test('requires navigator prop', () => {
    expect(chatScreen.props().navigator).toBeDefined();
  });
});
