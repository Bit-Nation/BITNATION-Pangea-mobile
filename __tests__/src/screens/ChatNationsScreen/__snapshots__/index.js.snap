// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatNationsScreen Rendering 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(ChatNationsScreen)
    fetchNations={[MockFunction]}
    navigator={
      Object {
        "pop": [MockFunction],
        "push": [MockFunction],
        "setButtons": [MockFunction],
        "setOnNavigatorEvent": [MockFunction],
      }
    }
    onSelectTab={[MockFunction]}
    openNation={[MockFunction]}
    startNationCreation={[MockFunction]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "fetchNations": [Function],
      "navigator": Object {
        "pop": [MockFunction],
        "push": [MockFunction],
        "setButtons": [MockFunction],
        "setOnNavigatorEvent": [MockFunction],
      },
      "onSelectTab": [Function],
      "openNation": [Function],
      "startNationCreation": [Function],
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "fetchNations": [Function],
        "navigator": Object {
          "pop": [MockFunction],
          "push": [MockFunction],
          "setButtons": [MockFunction],
          "setOnNavigatorEvent": [MockFunction],
        },
        "onSelectTab": [Function],
        "openNation": [Function],
        "startNationCreation": [Function],
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
